haystack/tools/from_function.py

```diff
@@ -0,0 +1,139 @@
+    >>>         'description': 'the unit for the temperature',
+    >>>         'default': 'Celsius',
+    >>>     },
+    >>>     }
+    >>> },
+    >>> function=<function get_weather at 0x7f7b3a8a9b80>)
+    ```
+
+    :param function:
+        The function to be converted into a Tool.
+        The function must include type hints for all parameters.
+        If a parameter is annotated using `typing.Annotated`, its metadata will be used as parameter description.
+    :param name:
+        The name of the Tool. If not provided, the name of the function will be used.
+    :param description:
+        The description of the Tool. If not provided, the docstring of the function will be used.
+        To intentionally leave the description empty, pass an empty string.
+
+    :returns:
+        The Tool created from the function.
+
+    :raises ValueError:
+        If any parameter of the function lacks a type hint.
+    :raises SchemaGenerationError:
+        If there is an error generating the JSON schema for the Tool.
+    """
+
+    tool_description = description if description is not None else (function.__doc__ or "")
+
+    signature = inspect.signature(function)
+
+    # collect fields (types and defaults) and descriptions from function parameters
+    fields: Dict[str, Any] = {}
+    descriptions = {}
+
+    for param_name, param in signature.parameters.items():
+        if param.annotation is param.empty:
+            raise ValueError(f"Function '{function.__name__}': parameter '{param_name}' does not have a type hint.")
+
+        # if the parameter has not a default value, Pydantic requires an Ellipsis (...)
+        # to explicitly indicate that the parameter is required
+        default = param.default if param.default is not param.empty else ...
+        fields[param_name] = (param.annotation, default)
+
+        if hasattr(param.annotation, "__metadata__"):
+            descriptions[param_name] = param.annotation.__metadata__[0]
+
+    # create Pydantic model and generate JSON schema
+    try:
+        model = create_model(function.__name__, **fields)
+        schema = model.model_json_schema()
+    except Exception as e:
+        raise SchemaGenerationError(f"Failed to create JSON schema for function '{function.__name__}'") from e
+
+    # we don't want to include title keywords in the schema, as they contain redundant information
+    # there is no programmatic way to prevent Pydantic from adding them, so we remove them later
+    # see https://github.com/pydantic/pydantic/discussions/8504
+    _remove_title_from_schema(schema)
+
+    # add parameters descriptions to the schema
+    for param_name, param_description in descriptions.items():
+        if param_name in schema["properties"]:
+            schema["properties"][param_name]["description"] = param_description
+
+    return Tool(name=name or function.__name__, description=tool_description, parameters=schema, function=function)
+
+
+def tool(function: Callable) -> Tool:
```

Comment:
the decorator essentially mirrors `create_tool_from_function`, but keeping them separate allows for potential future customization of the decorator.

Reply:
Nice, perhaps just mention in the pydocs of `create_tool_from_function` that one can simply use decorator instead of this function?

Reply:
done